### ISSUE TITLE
Materialization not working with DuckDB versions 1.1.x

### ISSUE THREAD

**Issue by mchlrch** [ontop/ontop#828]
### Description

I have set up a conversion from CSV files to RDF, using the latest available versions of DuckDB and its JDBC driver. Using the the ontop CLI, it does not materialize but stalls without error most of the time.

DuckDB versions 1.1.1 and 1.1.0 are both are not working (most of the time)
DuckDB version 1.0.0 is working OK

### Steps to Reproduce

Download DuckDB 1.1.1:
* https://github.com/duckdb/duckdb/releases/download/v1.1.1/duckdb_cli-linux-amd64.zip
* https://repo1.maven.org/maven2/org/duckdb/duckdb_jdbc/1.1.1/duckdb_jdbc-1.1.1.jar

**Expected behavior:** [What you expect to see]

(Using DuckDB 1.0.0) Materialization works as expected:

```
$ ./bin/ontop/ontop materialize -m ./src-gen/mapping-nova-zonenplan.r2rml.ttl -f ntriples -o ./target/foo.nt -p nova.properties
11:48:33.832 |-INFO  in i.u.i.o.a.r.impl.QuestQueryProcessor - Ontop has completed the setup and it is ready for query answering!
NR of TRIPLES: 2
Elapsed time to materialize: 181 {ms}
```

**Actual behavior:** [What actually happens]

(Using DuckDB 1.1.1) Seems stalled, nothing happens. No errors either

```
$ ./bin/ontop/ontop materialize -m ./src-gen/mapping-nova-zonenplan.r2rml.ttl -f ntriples -o ./target/foo.nt -p nova.properties
11:46:26.625 |-INFO  in i.u.i.o.a.r.impl.QuestQueryProcessor - Ontop has completed the setup and it is ready for query answering!
```

**Reproduces how often:** [Does it happen every time?]

Most of the time, but not always

### Attached material

Informations about the attachments (ontology, mappings and queries that are necessary to reproduce the problem)
Provide the log file (more information on how to obtain it in the guidelines)

### Versions

Which data source are you using, which JDBC driver and which versions (both the data source and the JDBC driver)

Which version of Ontop are you using and in which setting (Protege, SPARQL endpoint, Cli, Java API, ...)

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by bcogrel** [comment]
Hi @mchlrch, thanks for reporting the issue.

I didn't manage to reproduce it, things were working ok on my side with DuckDB 1.1.0 and 1.1.2.
Would it be possible for you to share a small example (data + mapping + properties) reproducing it?

On my side I tried with a SQL expression in the lens using the `read_parquet` table function and I had to enable the option `ontop.allowRetrievingBlackBoxViewMetadataFromDB=true`.

Best,
Benjamin

**Comment by rossbowen** [comment]
Hey, this may or may not be related (so happy to open another issue if not) but I came here because I'm also having trouble with the latest version of duckdb. Yasgui or querying the sparql endpoint results in a hanging behaviour. The logs do not show anything, even when set to `DEBUG`.

I'm running duckdb `v1.1.2 f680b7d08f`, the `duckdb_jdbc-1.1.1.jar` JDBC connector and ontop `5.2.0`.

As a reproducible example:

This creates a small duckdb database.

```python
# initialise.py

import duckdb
import pandas as pd

people_data = {
    'id': [1, 2, 3],
    'name': ['Alice', 'Bob', 'Charlie'],
    'age': [30, 25, 35]
}
people = pd.DataFrame(people_data)

conn = duckdb.connect("example.db")

conn.sql("CREATE TABLE people AS SELECT * FROM people")

conn.close()
```

This is my connection file for ontop.

```properties
# connection.properties

jdbc.url = jdbc:duckdb:example.db
jdbc.driver = org.duckdb.DuckDBDriver
```

I use `ontop bootstrap` to create mapping and ontology files.

```sh
ontop bootstrap \
    --ontology=ontology.xml \
    --mapping=mapping.obda \
    --properties=connection.properties \
    --base-iri=https://example.org/
```

Finally I start the sparql endpoint.

```sh
ontop endpoint \
    --ontology=ontology.xml \
    --mapping=mapping.obda \
    --properties=connection.properties \
    --cors-allowed-origins=http://yasgui.org
```

**Comment by canwaf** [comment]
Similar problem to @rossbowen here; however I've gone and attempted to materialize directly from the onto.obda file generated by Protege-Ontop, which is able to serialize using the GUI.

Running ontop version `5.2.0`, and `duckdb_jdbc-1.1.1.jar`.

```bash
ontop materialize -p rma-onto.properties \
    -m rma-onto.obda
    -f turtle
    -o test.ttl
```

From which it hangs after
```bash
09:57:32.853 |-INFO  in i.u.i.o.a.r.impl.QuestQueryProcessor - Ontop has completed the setup and it is ready for query answering!
```

**Comment by mchlrch** [comment]
Hi Benjamin

> I didn't manage to reproduce it, things were working ok on my side with DuckDB 1.1.0 and 1.1.2. Would it be possible for you to share a small example (data + mapping + properties) reproducing it?

Sure. Here's the reproducer: https://github.com/mchlrch/reproducer-ontop-828

With DuckDB 1.1.2 it sometimes works, but mostly stalls. Maybe it's some kind of timing/async issue.

I'm running this on Ubuntu 22.04, using OpendJDK 21

**Comment by LorenzBuehmann** [comment]
there is some kind of regression or even bug in the current JDBC code of DuckDB: https://github.com/duckdb/duckdb-java/issues/101
It does not accept a second connection which is odd ...

Debugging the code I can see Ontop is stuck when opening the second JDBC connection to handle another PPMapping.
It happens with all JDBC libs of version 1.1.x

It does not happen and work when going back to JDBC version 1.0.0

**Comment by bcogrel** [comment]
Thanks @mchlrch for the reproducer, very useful!

On my side, I always encounter the deadlock when running it in the command line.
Interestingly, on IntelliJ, the deadlock disappears when I run the debugger but always shows up when I run the test normally…

It definitely looks like some timing issue around accessing DuckDB DB files. I tried many configuration keys like setting DuckDB in read-only mode, but none helped so far.

As a workaround for keeping working with the latest version of DuckDB in your setting, I propose the following:
   - Instead of creating a DuckDB DB file, use a lens which will directly read the CSV file:
 ```sql
  {
    "relations": [
        {
            "name": [
                "zonen"
            ],
            "query": "SELECT * FROM read_csv('zonen.csv')",
            "type": "SQLLens"
        }
    ]
}
```
- Add the entry `ontop.allowRetrievingBlackBoxViewMetadataFromDB=true` to the properties file
- Replace `jdbc.url=jdbc:duckdb:my.duckdb` by `jdbc.url=jdbc:duckdb:`

That way you avoid any locking around the DB file.

**Comment by bcogrel** [comment]
Another workaround is to extract the DB metadata first (using `ontop extract-db-metadata`) and use that file for the materialization. This can save Ontop from opening a JDBC connection for extracting DB metadata and closing it just before opening another one for materializing.

My impression is that DuckDB's implementation of `Connection.close()` unblocks before the connection is actually completely closed.

**Comment by bcogrel** [comment]
Same issue with DuckDB 1.2.0

**Comment by bcogrel** [comment]
Good news: the problem disappears with DuckDB 1.3.0 ([jar](https://github.com/duckdb/duckdb-java/releases/tag/v1.3.0.0))!

It looks like the problem was https://github.com/duckdb/duckdb-java/issues/101, likely due to a statement not closed before the connection on the Ontop side. It would be good to find it.

**Comment by mchlrch** [comment]
> Good news: the problem disappears with DuckDB 1.3.0 ([jar](https://github.com/duckdb/duckdb-java/releases/tag/v1.3.0.0))!

Thanks for keeping an eye on this and sharing the good news 👍 

Soonish I will touch the originally affected pipeline and will switch to 1.3.0 then as well
